<launch>

  <arg name="GUI" default="true" />
  <arg name="PROCESS" default="true" />

  <include file="$(find mercury_ros)/launch/mask_rcnn_instance_segmentation.launch">
    <arg name="GUI" value="false" />
    <arg name="PROCESS" value="$(arg PROCESS)" />
  </include>

  <group ns="/camera/depth_registered">
    <node name="points_passthrough"
          pkg="jsk_topic_tools" type="passthrough"
          clear_params="true">
      <remap from="~input" to="points" />
      <rosparam subst_value="true">
        default_duration: 0
      </rosparam>
    </node>
  </group>

  <!-- for static scene -->
  <group ns="camera">
    <node name="render_voxel_grids"
          pkg="morefusion_ros" type="render_voxel_grids.py"
          clear_params="true" if="$(arg PROCESS)">
    </node>
    <node name="octomap_server"
          pkg="morefusion_ros" type="octomap_server"
          clear_params="true"
          output="screen" if="$(arg PROCESS)">
      <remap from="~render" to="render_voxel_grids/render" />
      <remap from="~input/camera_info" to="color/camera_info" />
      <remap from="~input/depth" to="aligned_depth_to_color/image_raw" />
      <remap from="~input/points" to="depth_registered/points_passthrough/output" />
      <remap from="~input/label_ins" to="mask_rcnn_instance_segmentation/output/label_ins" />
      <remap from="~input/class" to="mask_rcnn_instance_segmentation/output/class" />
      <rosparam subst_value="true">
        frame_id: panda_link0
        resolution: 0.01
        ground_as_noentry: true
      </rosparam>
    </node>
    <group ns="octomap_server">
      <node name="draw_instance_segmentation"
            pkg="morefusion_ros" type="draw_instance_segmentation.py"
            clear_params="true">
        <remap from="~input/rgb" to="/camera/color/image_rect_color_passthrough/output" />
        <remap from="~input/label_ins" to="output/label_tracked" />
        <remap from="~input/class" to="output/class" />
      </node>
    </group>
  </group>
  <node name="singleview_3d_pose_estimation_static"
        pkg="morefusion_ros" type="singleview_3d_pose_estimation.py"
        clear_params="true"
        output="screen" if="$(arg PROCESS)">
    <remap from="~input/camera_info" to="/camera/color/camera_info" />
    <remap from="~input/rgb" to="/camera/color/image_rect_color_passthrough/output" />
    <remap from="~input/depth" to="/camera/aligned_depth_to_color/image_raw" />
    <remap from="~input/label_ins" to="/camera/octomap_server/output/label_tracked" />
    <remap from="~input/class" to="/camera/octomap_server/output/class" />
    <remap from="~input/grids_noentry" to="/camera/octomap_server/output/grids_noentry" />
    <rosparam subst_value="true">
      with_occupancy: true
      icp: true
    </rosparam>
  </node>
  <node name="object_poses_to_markers_static"
        pkg="morefusion_ros" type="object_poses_to_markers.py"
        clear_params="true">
    <remap from="~input" to="singleview_3d_pose_estimation_static/output" />
  </node>
  <node name="object_mapping"
        pkg="morefusion_ros" type="object_mapping.py"
        clear_params="true" if="$(arg PROCESS)">
    <remap from="~input/poses" to="singleview_3d_pose_estimation_static/output" />
    <remap from="~input/grids" to="/camera/octomap_server/output/grids" />
    <rosparam subst_value="true">
      frame_id: panda_link0
      n_votes: 2
    </rosparam>
  </node>
  <group ns="object_mapping">
    <node name="object_poses_to_markers"
          pkg="morefusion_ros" type="object_poses_to_markers.py"
          clear_params="true">
      <remap from="~input" to="output/poses" />
    </node>
    <node name="grids_to_markers"
          pkg="morefusion_ros" type="voxel_grids_to_markers.py"
          clear_params="true">
      <remap from="~input" to="output/grids" />
      <rosparam subst_value="true">
        show_bbox: false
      </rosparam>
    </node>
    <node name="grids_to_centroids"
          pkg="mercury_ros" type="voxel_grids_to_centroids.py"
          clear_params="true">
      <remap from="~input" to="output/grids" />
    </node>
  </group>

  <!-- for dynamic scene -->
  <node name="singleview_3d_pose_estimation"
        pkg="morefusion_ros" type="singleview_3d_pose_estimation.py"
        clear_params="true"
        output="screen" if="$(arg PROCESS)">
    <remap from="~input/camera_info" to="/camera/color/camera_info" />
    <remap from="~input/rgb" to="/camera/color/image_rect_color_passthrough/output" />
    <remap from="~input/depth" to="/camera/aligned_depth_to_color/image_raw" />
    <remap from="~input/label_ins" to="/camera/mask_rcnn_instance_segmentation/output/label_ins" />
    <remap from="~input/class" to="/camera/mask_rcnn_instance_segmentation/output/class" />
    <rosparam subst_value="true">
      with_occupancy: false
      icp: true
    </rosparam>
  </node>
  <node name="object_poses_to_markers"
        pkg="morefusion_ros" type="object_poses_to_markers.py"
        clear_params="true">
    <remap from="~input" to="singleview_3d_pose_estimation/output" />
  </node>

  <node name="republish_tf_static_for_rosbag"
        pkg="mercury_ros" type="republish_tf_static_for_rosbag.py"
        clear_params="true" if="$(arg PROCESS)" />

  <group if="$(arg GUI)">
    <node name="rviz"
          pkg="rviz" type="rviz"
          args="-d $(find mercury_ros)/rviz/morefusion.rviz"
          clear_params="true">
    </node>
  </group>

</launch>
